src/workqueue/workqueue.c

```diff
@@ -0,0 +1,849 @@
+ * thread that dequeues it to suspend itself.
+ */
+
+
+/////////////////////////////////
+// Interface between core workqueue and scheduling management subsystem
+/////////////////////////////////
+
+// The number of runnable worker threads as observed by
+// the scheduling management subsystem.
+// TODO: This may be moved "down" to user-level scheduler
+//       if it is not easy to maintain from the kext.
+static volatile int32_t runnable_workers;
+
+// The desired minimum number of runnable worker threads
+static int32_t target_runnable_workers;
+
+// The desired maximum number of runnable worker threads
+static int32_t target_max_runnable_workers;
+
+// Limit on the total number of worker threads that can be created.
+static int32_t max_total_workers;
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+// Are we using user-level or kext based management?
+static bool kext_active;
+#endif
+
+// Called on startup of the workqueue system
+static void _dispatch_workqueue_management_init(void);
+
+// Called by worker thread when it starts executing
+static int _dispatch_worker_register(void);
+
+// Called by worker thread just before it exits
+static void _dispatch_worker_unregister(void);
+
+// Called when new work is added to the system
+static void _dispatch_work_added(void);
+
+// Called by a worker thread when either it has found no work to do
+// or it has been command to wait by the management subsystem.
+// This call is expected to block until the management subsystem
+// thinks there may be work for the thread to do.
+static void _dispatch_worker_wait(bool was_ordered_to_wait);
+
+// Called by the management system when it wants to inject a
+// workitem that will case a worker thread to wait
+static void _dispatch_worker_should_wait(void);
+
+// Called by the management system when it wants to inject a
+// workitem that will case a worker thread to terminate
+static void _dispatch_worker_should_exit(void);
+
+// Called by the management system when it wants to
+// create an additional worker thread.
+static void _dispatch_worker_should_start(void);
+
+#if DISPATCH_ENABLE_PWQ_KEXT
+//////////////////////////////////////////////////////////////////
+// versions of above functions required for the kernel extension
+// temporary to isolate kext code to separate file.
+//////////////////////////////////////////////////////////////////
+
+int _dispatch_workqueue_management_init_kext(void);
+void _dispatch_worker_register_kext(void);
+void _dispatch_worker_unregister_kext(void);
+void _dispatch_work_added_kext(void);
+void _dispatch_worker_wait_kext(void);
+void* dispatch_thread_monitor_main_kext(void *context DISPATCH_UNUSED);
+
+#endif // DISPATCH_ENABLE_PWQ_KEXT
+
+/////////////////////////////////
+// Workqueue internal data structures and functions
+/////////////////////////////////
+
+// Debug messages for threadpool operations
+static bool threadpool_verbose;
+#define verbose_printf(fmt,...) do {								\
```

Comment:
We tend to instead have a `#define DISPATCH_WQ_DEBUG` that is set either inidividually or if `DISPATCH_DEBUG` is set so that it is compiled out by default.